atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DefaultShardProgressUpdater.java

```diff
@@ -0,0 +1,160 @@
+                Bucket.of(shardAndStrategy, bucketProbeResult.endExclusive()));
+        sweepQueueProgressUpdater.progressTo(shardAndStrategy, bucketProbeResult.knownSweepProgress());
+    }
+
+    /**
+     * Returns a {@link BucketProbeResult} indicating a prefix of cells that have been swept successfully, starting
+     * from a given point in time. It is assumed that all buckets before searchStart have been swept successfully;
+     * if this is not the case, behaviour is undefined.
+     */
+    private BucketProbeResult findCompletedBuckets(ShardAndStrategy shardAndStrategy, long searchStart) {
+        for (long offset = 0; offset < MAX_BUCKETS_TO_CHECK_PER_ITERATION; offset++) {
+            long currentBucket = searchStart + offset;
+            Optional<BucketProgress> bucketProgress =
+                    bucketProgressStore.getBucketProgress(Bucket.of(shardAndStrategy, currentBucket));
+            if (bucketProgress.isPresent()) {
+                BucketProgress presentBucketProgress = bucketProgress.get();
+                TimestampRange requiredRange = getTimestampRangeRecord(currentBucket);
+                if (presentBucketProgress.timestampProgress()
+                        != requiredRange.endExclusive() - requiredRange.startInclusive() - 1) {
+                    // Bucket still has progress to go, so we can stop here.
+                    return BucketProbeResult.builder()
+                            .endExclusive(currentBucket)
+                            .knownSweepProgress(
+                                    requiredRange.startInclusive() + presentBucketProgress.timestampProgress())
+                            .build();
+                } else {
+                    // Bucket fully processed, keep going.
+                    if (offset == MAX_BUCKETS_TO_CHECK_PER_ITERATION - 1) {
+                        // We finished the maximum number of buckets to check, and all were completed.
+                        return BucketProbeResult.builder()
+                                .endExclusive(currentBucket + 1)
+                                .knownSweepProgress(requiredRange.endExclusive() + 1)
+                                .build();
+                    }
+                }
+            } else {
+                // No progress; we're ahead of the read pointer, so interpret as unstarted.
+                return BucketProbeResult.builder()
+                        .endExclusive(currentBucket)
+                        .knownSweepProgress(
+                                getTimestampRangeRecord(currentBucket).startInclusive() - 1L)
+                        .build();
+            }
+        }
+        throw new SafeIllegalStateException("Didn't expect to get here");
+    }
+
+    private TimestampRange getTimestampRangeRecord(long queriedBucket) {
+        try {
+            return recordsTable.getTimestampRangeRecord(queriedBucket);
+        } catch (NoSuchElementException exception) {
```

Comment:
note: Either make the change to NoSuchElementException in this PR, or I can make it in a separate one - it currently does not throw this

Reply:
https://github.com/palantir/atlasdb/pull/7370

Reply:
(thanks!)